core/src/com/unciv/ui/utils/UnitGroup.kt

```diff
@@ -4,30 +4,103 @@ import com.badlogic.gdx.graphics.Color
 import com.badlogic.gdx.scenes.scene2d.Group
 import com.badlogic.gdx.scenes.scene2d.actions.Actions
 import com.badlogic.gdx.scenes.scene2d.actions.RepeatAction
+import com.badlogic.gdx.scenes.scene2d.actions.SequenceAction
 import com.badlogic.gdx.scenes.scene2d.ui.Image
+import com.badlogic.gdx.utils.Align
 import com.unciv.UncivGame
 import com.unciv.logic.map.MapUnit
 import com.unciv.ui.images.ImageGetter
 import com.unciv.ui.utils.extensions.center
+import com.unciv.ui.utils.extensions.centerX
 import com.unciv.ui.utils.extensions.surroundWithCircle
+import com.unciv.ui.utils.extensions.toGroup
 
 class UnitGroup(val unit: MapUnit, val size: Float): Group() {
-    var blackSpinningCircle: Image? = null
     var actionGroup :Group? = null
     val unitBaseImage = ImageGetter.getUnitIcon(unit.name, unit.civInfo.nation.getInnerColor())
-        .apply { setSize(size * 0.75f, size * 0.75f) }
-    var background: Image? = null
+        .apply {
+            if (unit.isCivilian())
+                setSize(size * 0.60f, size * 0.60f)
+            else
+                setSize(size * 0.75f, size * 0.75f) }
+    var flagSelection: Image? = null
+    var flagHalo: Image? = null
+    var flagBg: Group? = null
 
     init {
-        background = getBackgroundImageForUnit()
-        background?.apply {
-            this.color = unit.civInfo.nation.getOuterColor()
-            this.color.a = UncivGame.Current.settings.unitIconOpacity
-            setSize(size, size)
+
+        val bgW = size
+        var bgH = size
+
+        if (unit.isFortified())
+            bgH *= 1.1f
+        else if (unit.isEmbarked())
+            bgH *= 1.1f
+        else if (unit.isCivilian())
+            bgH *= 1.2f
```

Comment:
This is hardcoded and doesn't leave modders much choice to change icons and have everything work
Please calculate according to icon width / icon height
(Also please rename to backgroundWidth and backgroundHeight, we try to keep the code as immediately understandable as possible 😉 )